promql/engine.go

```diff
@@ -0,0 +1,1206 @@
+func (r Result) Vector() (Vector, error) {
+	if r.Err != nil {
+		return nil, r.Err
+	}
+	v, ok := r.Value.(Vector)
+	if !ok {
+		return nil, fmt.Errorf("query result is not a vector")
+	}
+	return v, nil
+}
+
+// Matrix returns a matrix. An error is returned if
+// the result was an error of the result value is not a matrix.
+func (r Result) Matrix() (Matrix, error) {
+	if r.Err != nil {
+		return nil, r.Err
+	}
+	v, ok := r.Value.(Matrix)
+	if !ok {
+		return nil, fmt.Errorf("query result is not a matrix")
+	}
+	return v, nil
+}
+
+// Scalar returns a scalar value. An error is returned if
+// the result was an error of the result value is not a scalar.
+func (r Result) Scalar() (*Scalar, error) {
+	if r.Err != nil {
+		return nil, r.Err
+	}
+	v, ok := r.Value.(*Scalar)
+	if !ok {
+		return nil, fmt.Errorf("query result is not a scalar")
+	}
+	return v, nil
+}
+
+func (res Result) String() string {
+	if res.Err != nil {
+		return res.Err.Error()
+	}
+	if res.Value == nil {
+		return ""
+	}
+	return res.Value.String()
+}
+
+// A Query is derived from an a raw query string and can be run against an engine
+// it is associated with.
+type Query interface {
+	// Exec starts evaluating the query and returns immediately.
+	// The returned channel blocks until the evaluation is done and receives
+	// false if an error occurred, true otherwise.
+	Exec() (done <-chan bool)
+	// Result returns the result of the evaluated query. It blocks until the
+	// result is available.
+	Result() Result
+	// Statements returns the parsed statements of the query.
+	Statements() Statements
+	// Stats returns statistics about the lifetime of the query.
+	Stats() *stats.TimerGroup
+	// Cancel signals that a running query execution should be aborted.
+	Cancel()
+}
+
+// query implements the Query interface.
+type query struct {
+	q      string     // The original query string.
+	stmts  Statements // Statements of the parsed query.
+	result Result     // Result of the evaluated query.
+	done   chan bool
+	stats  *stats.TimerGroup
+	cancel func()
+
+	ng *Engine // The engine against which the query is executed.
+}
+
+// Statements implements the Query interface.
+func (q *query) Statements() Statements {
+	return q.stmts
+}
+
+// Stats implements the Query interface.
+func (q *query) Stats() *stats.TimerGroup {
+	return q.stats
+}
+
+// Cancel implements the Query interface.
+func (q *query) Cancel() {
+	if q.cancel != nil {
+		q.cancel()
+	}
+}
+
+// Exec implements the Query interface.
+func (q *query) Exec() <-chan bool {
+	ctx := q.ng.setupQuery(q)
+	go func() {
+		q.result = q.ng.exec(ctx, q)
+		success := q.result.Err == nil
+		q.done <- success
+		q.done <- success
+	}()
+	return q.done
+}
+
+// Result retrieves the result from an executed query. It blocks as long as
+// the execution is running.
+func (q *query) Result() Result {
+	if q.done == nil {
+		return Result{Err: errors.New("query execution not started")}
```

Comment:
`q.done` is constructed in `NewQuery`, so it'll never be `nil` unless there's a programming error, right?